src/GraphQL/Validation/ValidationContext.cs

```diff
@@ -204,31 +200,277 @@ public List<FragmentDefinition> GetRecursivelyReferencedFragments(Operation oper
+        {
+            if ((variableDefinitions?.List?.Count ?? 0) == 0)
+            {
+                return Variables.None;
+            }
+
+            var variables = new Variables(variableDefinitions.List.Count);
+
+            if (variableDefinitions != null)
+            {
+                foreach (var variableDef in variableDefinitions.List)
+                {
+                    // find the IGraphType instance for the variable type
+                    var graphType = variableDef.Type.GraphTypeFromType(schema);
+
+                    if (graphType == null)
+                    {
+                        ReportError(new InvalidVariableError(this, variableDef, variableDef.Name, $"Variable has unknown type '{variableDef.Type.Name()}'"));
+                        continue;
+                    }
+
+                    // create a new variable object
+                    var variable = new Variable
+                    {
+                        Name = variableDef.Name
+                    };
+
+                    // attempt to retrieve the variable value from the inputs
+                    if (inputs.TryGetValue(variableDef.Name, out var variableValue))
+                    {
+                        // parse the variable via ParseValue (for scalars) and ParseDictionary (for objects) as applicable
+                        try
+                        {
+                            variable.Value = GetVariableValue(graphType, variableDef, variableValue, visitor);
+                        }
+                        catch (ValidationError error)
+                        {
+                            ReportError(error);
+                            continue;
+                        }
+                    }
+                    else if (variableDef.DefaultValue != null)
+                    {
+                        // if the variable was not specified in the inputs, and a default literal value was specified, use the specified default variable value
+
+                        // parse the variable literal via ParseLiteral (for scalars) and ParseDictionary (for objects) as applicable
+                        try
+                        {
+                            variable.Value = ExecutionHelper.CoerceValue(graphType, variableDef.DefaultValue, variables, null).Value;
+                        }
+                        catch (Exception ex)
+                        {
+                            ReportError(new InvalidVariableError(this, variableDef, variableDef.Name, "Error coercing default value.", ex));
+                            continue;
+                        }
+                        variable.IsDefault = true;
+                    }
+                    else if (graphType is NonNullGraphType)
+                    {
+                        ReportError(new InvalidVariableError(this, variableDef, variable.Name, "Received a null input for a non-null variable."));
```

Comment:
Is this error more specifically that no variable value was provided for a non-null variable?
```suggestion
                        ReportError(new InvalidVariableError(this, variableDef, variable.Name, "No value provided for a non-null variable."));
```

Reply:
👍 

The same for 
>  // RULE: If no default value is provided and the input object field’s type is non‐null,
                        // an error should be thrown.
                        throw new InvalidVariableError(this, variableDef, childFieldVariableName, "Received a null input for a non-null variable.");


?

Reply:
Done for both.

Reply:
Yes, correct.